Upload frequently aborted on transient 503 Service Unavailable for multi-module

I hit frequent 503 Service Unavailable error for multi-module deployments.

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [ ] Example that reproduces the problem (link to git repository is ideal)
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Dispatch a multi-module deployment
2. JReleaser aborted on 503 at `MavenCentral.artifactExists`
3. 2nd deployment succeeded 
4. But the next deployment aborted on 503 at `MavenCentral.artifactExists` again (and its 2nd succeeded again)


### Expected Behaviour

- Retry on 503 

### Actual Behaviour

JReleaser aborts on 503:

```
[INFO]      [mavenCentral]  - net.exoego.uika-uika-0.0.2-bundle.zip
[INFO]  Writing output properties to out/jreleaser/output.properties
Error:  JReleaser failed after 01:03 m
Error:  Look for more details at /home/runner/work/uika/uika/out/jreleaser/trace.log
org.jreleaser.sdk.commons.RestAPIException: 503: Request replied with: 503 reason: Service Unavailable
	at org.jreleaser.sdk.commons.ClientUtils.head(ClientUtils.java:476)
	at org.jreleaser.sdk.mavencentral.MavenCentral.artifactExists(MavenCentral.java:115)
	at org.jreleaser.sdk.mavencentral.MavenCentralMavenDeployer.uploadArtifacts(MavenCentralMavenDeployer.java:206)
	at org.jreleaser.sdk.mavencentral.MavenCentralMavenDeployer.deploy(MavenCentralMavenDeployer.java:106)
	at org.jreleaser.engine.deploy.maven.ProjectMavenDeployer.deploy(ProjectMavenDeployer.java:55)
	at org.jreleaser.engine.deploy.maven.MavenDeployers.deploy(MavenDeployers.java:167)
	at org.jreleaser.engine.deploy.maven.MavenDeployers.doDeploy(MavenDeployers.java:150)
	at org.jreleaser.engine.deploy.maven.MavenDeployers.deploy(MavenDeployers.java:59)
	at org.jreleaser.workflow.DeployWorkflowItem.doInvoke(DeployWorkflowItem.java:35)
	at org.jreleaser.workflow.AbstractWorkflowItem.lambda$invoke$0(AbstractWorkflowItem.java:43)
	at org.jreleaser.engine.hooks.HookExecutor.execute(HookExecutor.java:83)
	at org.jreleaser.workflow.AbstractWorkflowItem.invoke(AbstractWorkflowItem.java:43)
	at org.jreleaser.workflow.WorkflowImpl.doExecute(WorkflowImpl.java:130)
	at org.jreleaser.workflow.WorkflowImpl.execute(WorkflowImpl.java:54)
	at org.jreleaser.cli.Release.doExecute(Release.java:387)
	at org.jreleaser.cli.AbstractModelCommand.execute(AbstractModelCommand.java:103)
	at org.jreleaser.cli.Release.execute(Release.java:311)
	at org.jreleaser.cli.AbstractCommand.call(AbstractCommand.java:52)
	at org.jreleaser.cli.AbstractCommand.call(AbstractCommand.java:37)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2031)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2469)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2423)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2425)
	at picocli.CommandLine.execute(CommandLine.java:2174)
	at org.jreleaser.cli.Main.execute(Main.java:98)
	at org.jreleaser.cli.Main.run(Main.java:81)
	at org.jreleaser.cli.Main.main(Main.java:70)
Error: Process completed with exit code 1.
```


### Environment Information

- **Operating System**: Ubuntu (GitHub Actions default runner)
- **JReleaser Version:** 1.25.0
- **JDK Version:** 21

Please paste stacktraces from `[out|build|target]/jreleaser/trace.log`.

--- a/api/jreleaser-resource-bundle/src/main/resources/org/jreleaser/bundle/Messages.properties
+++ b/api/jreleaser-resource-bundle/src/main/resources/org/jreleaser/bundle/Messages.properties
@@ -881,6 +881,7 @@ maven.central.upload.bundle             = Uploading {}
 maven.central.bundled.uploaded          = Bundle {} uploaded as deployment {}
 maven.central.publish.deployment        = Publishing deployment {}
 maven.central.deployment.status         = querying status for deployment {}
+maven.central.published.check           = checking if {}:{}:{} is published
 WARN_maven_central_publication_timeout  = Deployment timeout exceeded. However, the remote operation may still be successful. Please log into Maven Central and check the status for deployment {}
 ERROR_maven_central_find_deployment     = Could not find a deployment for {}
 ERROR_maven_central_create_bundle       = Could not create deployment bundle {}

--- a/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
@@ -108,12 +108,12 @@ public class MavenCentral {
     }
 
     public boolean isPublished(String groupId, String artifactId, String version) throws MavenCentralException {
-        Optional<PublishedStatus> status = wrap(() -> {
+        PublishedStatus status = wrap(() -> {
             context.getLogger().debug(RB.$("maven.central.published.check"), groupId, artifactId, version);
-            return Optional.ofNullable(api.published(groupId, artifactId, version));
+            return api.published(groupId, artifactId, version);
         });
 
-        return null != status && status.map(PublishedStatus::isPublished).orElse(false);
+        return null != status && status.isPublished();
     }
 
     public Optional<Deployment> status(String deploymentId) throws MavenCentralException {

--- a/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
@@ -36,13 +36,12 @@ import org.jreleaser.bundle.RB;
 import org.jreleaser.logging.JReleaserLogger;
 import org.jreleaser.model.Http;
 import org.jreleaser.model.api.JReleaserContext;
-import org.jreleaser.model.spi.deploy.maven.Deployable;
-import org.jreleaser.mustache.Templates;
 import org.jreleaser.sdk.commons.ClientUtils;
 import org.jreleaser.sdk.commons.feign.TokenAuthRequestInterceptor;
 import org.jreleaser.sdk.mavencentral.api.Deployment;
 import org.jreleaser.sdk.mavencentral.api.MavenCentralAPI;
 import org.jreleaser.sdk.mavencentral.api.MavenCentralAPIException;
+import org.jreleaser.sdk.mavencentral.api.PublishedStatus;
 import org.jreleaser.sdk.mavencentral.api.State;
 
 import java.io.IOException;
@@ -62,7 +61,6 @@ import java.util.concurrent.Callable;
 import static java.lang.System.lineSeparator;
 import static java.util.Objects.requireNonNull;
 import static org.jreleaser.util.IoUtils.newInputStreamReader;
-import static org.jreleaser.util.StringUtils.isNotBlank;
 import static org.jreleaser.util.StringUtils.requireNonBlank;
 
 /**
@@ -109,18 +107,13 @@ public class MavenCentral {
             .target(MavenCentralAPI.class, apiHost);
     }
 
-    public boolean artifactExists(Deployable deployable, String verifyUrl) {
-        if (isNotBlank(verifyUrl)) {
-            verifyUrl = Templates.resolveTemplate(context.getLogger(), verifyUrl, deployable.props());
-            if (ClientUtils.head(context.getLogger(), verifyUrl, connectTimeout, readTimeout)) {
-                context.getLogger().warn(" ! " + RB.$("nexus.deploy.artifact.exists",
-                    deployable.getDeployPath(),
-                    deployable.getLocalPath().getFileName().toString()));
-                return true;
-            }
-        }
+    public boolean isPublished(String groupId, String artifactId, String version) throws MavenCentralException {
+        Optional<PublishedStatus> status = wrap(() -> {
+            context.getLogger().debug(RB.$("maven.central.published.check"), groupId, artifactId, version);
+            return Optional.ofNullable(api.published(groupId, artifactId, version));
+        });
 
-        return false;
+        return null != status && status.map(PublishedStatus::isPublished).orElse(false);
     }
 
     public Optional<Deployment> status(String deploymentId) throws MavenCentralException {

--- a/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
@@ -251,9 +251,14 @@ public class MavenCentral {
             RetryPolicy<R> policy = RetryPolicy.<R>builder()
                 .handle(IllegalStateException.class)
                 .handleIf(exception -> {
-                    if (exception instanceof MavenCentralAPIException) {
-                        MavenCentralAPIException mavenCentralException = (MavenCentralAPIException) exception;
-                        return !mavenCentralException.isUnauthorized();
+                    for (Throwable cause = exception; null != cause; cause = cause.getCause()) {
+                        if (cause instanceof MavenCentralAPIException) {
+                            MavenCentralAPIException mavenCentralException = (MavenCentralAPIException) cause;
+                            return !mavenCentralException.isUnauthorized();
+                        }
+                        if (cause instanceof RetryableException) {
+                            return true;
+                        }
                     }
 
                     return false;

--- a/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
@@ -113,7 +113,7 @@ public class MavenCentral {
             return api.published(groupId, artifactId, version);
         });
 
-        return null != status && status.isPublished();
+        return status.isPublished();
     }
 
     public Optional<Deployment> status(String deploymentId) throws MavenCentralException {

--- a/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentral.java
@@ -287,7 +287,7 @@ public class MavenCentral {
                 return exception;
             }
 
-            if (response.status() >= 500) {
+            if (response.status() >= 500 || 429 == response.status()) {
                 logger.trace(response.request().httpMethod() + " " + response.request().url());
                 logger.trace(response.status() + " " + response.reason());
                 if (null != response.body() && null != response.body().length() && response.body().length() > 0) {

--- a/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentralMavenDeployer.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/MavenCentralMavenDeployer.java
@@ -201,15 +201,26 @@ public class MavenCentralMavenDeployer extends AbstractMavenDeployer<org.jreleas
     private Optional<String> uploadArtifacts(MavenCentral mavenCentral, Path bundleZip) throws DeployException {
         context.getLogger().info(" - {}", bundleZip.getFileName());
 
-        boolean success = true;
-        for (Deployable deployable : collectDeployableArtifacts()) {
-            if (mavenCentral.artifactExists(deployable, context.getModel().getProject().isSnapshot() ? null : getDeployer().getVerifyUrl())) {
-                success = false;
+        if (!context.getModel().getProject().isSnapshot() && !context.isDryrun()) {
+            boolean success = true;
+            for (Deployable deployable : collectDeployableArtifacts()) {
+                if (!deployable.isPom()) continue;
+                try {
+                    if (mavenCentral.isPublished(deployable.getGroupId(), deployable.getArtifactId(), deployable.getVersion())) {
+                        context.getLogger().warn(" ! " + RB.$("nexus.deploy.artifact.exists",
+                            deployable.getDeployPath(),
+                            deployable.getLocalPath().getFileName().toString()));
+                        success = false;
+                    }
+                } catch (MavenCentralException e) {
+                    context.getLogger().trace(e);
+                    throw new DeployException(RB.$("ERROR_unexpected_error"), e);
+                }
             }
-        }
 
-        if (!success) {
-            throw new DeployException(RB.$("ERROR_nexus_deploy_artifacts"));
+            if (!success) {
+                throw new DeployException(RB.$("ERROR_nexus_deploy_artifacts"));
+            }
         }
 
         if (!context.isDryrun()) {

--- a/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/api/MavenCentralAPI.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/api/MavenCentralAPI.java
@@ -47,7 +47,7 @@ public interface MavenCentralAPI {
     @Headers("Content-Type: application/json")
     Deployment status(@QueryMap Map<String, Object> params);
 
-    @RequestLine("GET /published")
+    @RequestLine("GET /published?namespace={namespace}&name={name}&version={version}")
     @Headers("Content-Type: application/json")
-    Deployment published(@Param("namespace") String namespace, @Param("name") String name, @Param("version") String version);
+    PublishedStatus published(@Param("namespace") String namespace, @Param("name") String name, @Param("version") String version);
 }

--- a/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/api/PublishedStatus.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/main/java/org/jreleaser/sdk/mavencentral/api/PublishedStatus.java
@@ -1,0 +1,37 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright 2020-2026 The JReleaser authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jreleaser.sdk.mavencentral.api;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+/**
+ * @author Andres Almiray
+ * @since 1.26.0
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class PublishedStatus {
+    private boolean published;
+
+    public boolean isPublished() {
+        return published;
+    }
+
+    public void setPublished(boolean published) {
+        this.published = published;
+    }
+}

--- a/sdks/jreleaser-mavencentral-java-sdk/src/test/java/org/jreleaser/sdk/mavencentral/MavenCentralErrorDecoderTest.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/test/java/org/jreleaser/sdk/mavencentral/MavenCentralErrorDecoderTest.java
@@ -77,6 +77,15 @@ class MavenCentralErrorDecoderTest {
 
 
     @Test
+    void returnsRetryableExceptionFor429() {
+        MavenCentral.MavenCentralErrorDecoder decoder = new MavenCentral.MavenCentralErrorDecoder(logger, "testDeployer");
+        Response response = buildResponse(429, "Too Many Requests");
+        Exception ex = decoder.decode("method", response);
+        assertThat(ex).isInstanceOf(RetryableException.class);
+    }
+
+
+    @Test
     void returnsDefaultExceptionForOtherStatus() {
         MavenCentral.MavenCentralErrorDecoder decoder = new MavenCentral.MavenCentralErrorDecoder(logger, "testDeployer");
         Response response = buildResponse(404, "Not Found");

--- a/sdks/jreleaser-mavencentral-java-sdk/src/test/java/org/jreleaser/sdk/mavencentral/MavenCentralRetryTest.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/test/java/org/jreleaser/sdk/mavencentral/MavenCentralRetryTest.java
@@ -1,0 +1,91 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright 2020-2026 The JReleaser authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jreleaser.sdk.mavencentral;
+
+import dev.failsafe.FailsafeException;
+import feign.Request;
+import feign.RequestTemplate;
+import feign.RetryableException;
+import org.jreleaser.logging.JReleaserLogger;
+import org.jreleaser.logging.SimpleJReleaserLoggerAdapter;
+import org.jreleaser.sdk.mavencentral.api.MavenCentralAPIException;
+import org.junit.jupiter.api.Test;
+
+import java.util.Collections;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class MavenCentralRetryTest {
+    private final JReleaserLogger logger = new SimpleJReleaserLoggerAdapter();
+
+    @Test
+    void retrierRetriesWrappedTransientFailures() {
+        MavenCentral.Retrier retrier = new MavenCentral.Retrier(logger, 1, 2);
+        AtomicInteger attempts = new AtomicInteger();
+
+        String result = retrier.retry(r -> false, () -> {
+            if (attempts.incrementAndGet() == 1) {
+                throw new MavenCentralException("transient failure", retryable503());
+            }
+            return "ok";
+        });
+
+        assertThat(result).isEqualTo("ok");
+        assertThat(attempts.get()).isEqualTo(2);
+    }
+
+    @Test
+    void retrierRetriesWrappedApiExceptions() {
+        MavenCentral.Retrier retrier = new MavenCentral.Retrier(logger, 1, 2);
+        AtomicInteger attempts = new AtomicInteger();
+
+        String result = retrier.retry(r -> false, () -> {
+            if (attempts.incrementAndGet() == 1) {
+                throw new MavenCentralException("transient failure",
+                    new MavenCentralAPIException(503, "Service Unavailable"));
+            }
+            return "ok";
+        });
+
+        assertThat(result).isEqualTo("ok");
+        assertThat(attempts.get()).isEqualTo(2);
+    }
+
+    @Test
+    void retrierDoesNotRetryUnauthorized() {
+        MavenCentral.Retrier retrier = new MavenCentral.Retrier(logger, 1, 2);
+        AtomicInteger attempts = new AtomicInteger();
+
+        assertThatThrownBy(() -> retrier.retry(r -> false, () -> {
+            attempts.incrementAndGet();
+            throw new MavenCentralException("fatal failure",
+                new MavenCentralAPIException(401, "Unauthorized"));
+        })).isInstanceOf(FailsafeException.class)
+            .hasRootCauseInstanceOf(MavenCentralAPIException.class);
+
+        assertThat(attempts.get()).isEqualTo(1);
+    }
+
+    private RetryableException retryable503() {
+        Request request = Request.create(Request.HttpMethod.HEAD, "http://localhost", Collections.emptyMap(), null,
+            new RequestTemplate());
+        return new RetryableException(503, "Service Unavailable", Request.HttpMethod.HEAD, (Long) null, request);
+    }
+}

--- a/sdks/jreleaser-mavencentral-java-sdk/src/test/java/org/jreleaser/sdk/mavencentral/MavenCentralRetryTest.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/test/java/org/jreleaser/sdk/mavencentral/MavenCentralRetryTest.java
@@ -84,8 +84,7 @@ class MavenCentralRetryTest {
     }
 
     private RetryableException retryable503() {
-        Request request = Request.create(Request.HttpMethod.HEAD, "http://localhost", Collections.emptyMap(), null,
-            new RequestTemplate());
-        return new RetryableException(503, "Service Unavailable", Request.HttpMethod.HEAD, (Long) null, request);
+        return new RetryableException(503, "Service Unavailable", Request.HttpMethod.HEAD, (Long) null,
+            Request.create(Request.HttpMethod.HEAD, "http://localhost", Collections.emptyMap(), null, new RequestTemplate()));
     }
 }

--- a/sdks/jreleaser-mavencentral-java-sdk/src/test/java/org/jreleaser/sdk/mavencentral/api/PublishedStatusTest.java
+++ b/sdks/jreleaser-mavencentral-java-sdk/src/test/java/org/jreleaser/sdk/mavencentral/api/PublishedStatusTest.java
@@ -1,0 +1,34 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright 2020-2026 The JReleaser authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jreleaser.sdk.mavencentral.api;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class PublishedStatusTest {
+    @Test
+    void deserializesPublishedResponse() throws Exception {
+        ObjectMapper mapper = new ObjectMapper();
+
+        assertThat(mapper.readValue("{\"published\":true}", PublishedStatus.class).isPublished()).isTrue();
+        assertThat(mapper.readValue("{\"published\":false}", PublishedStatus.class).isPublished()).isFalse();
+        assertThat(mapper.readValue("{\"published\":true,\"unknown\":1}", PublishedStatus.class).isPublished()).isTrue();
+    }
+}

